Design.md

```diff
@@ -0,0 +1,55 @@
+
+## Runner
+
+The jobs themselves are normal child processes of the runner. The runner configures CGroups and Linux namespaces for children to limit resource usage and isolate the them from the other processes. It also handles output recording and forwarding. All data is stored in-memory. In case of the service process termination, all jobs and their status are lost.
+
+## Communication and auth
+
+Clients communicate via gRPC with a simple protocol: One service with four endpoints. The communication is secured [RusTLS](https://github.com/ctz/rustls) which by design supports only modern, secure cipher suites. [Tonic](https://github.com/hyperium/tonic) is used to serve gRPC and almost automatically handles protocol buffers, encryption and related concerns.
+
+Authentication is implemented with mTLS. Server and client have different CA roots, which they are expected to exchange in a secure way. (Scenario-wise: the client CA is operated by the same organization that hosts job-runner). The server identifies each client with it's CommonName (CN) field of the certificate. The client CA only issues certificates with CN values that begin with `client_` to developers that are allowed to access the API. Authorization is done by checking for this property; clients with other CN prefixes are not allowed to execute any operations. Moreover, all running jobs are bound to the CN of the client calling `Start`, and only a client with that certificate is allowed to execute operations for that job.
+
+A more complete system should also support certificate revocation lists (CRL), so that individual certificates can be revoked in case they are compromised. Other useful feature would be having more fine-grained access control. This could be done per certificate, or by using API keys in addition to the certificates. These could then be restricted to only specific endpoints, executables, and resource limits.
+
+## Endpoints
+
+### Start
+
+Starts a new job by spawning a process from given executable path and arguments. Returns a unique job id, that is used to specify the target job for other endpoints. If the executable is not found or cannot be executed, immediately returns an error.
+
+No security checks are applied to the program and arguments. If the client wishes to remove some files or start a fork bomb, it's allowed to do so.
+
+### Stop
+
+Cancels a job by sending `SIGKILL`. Stop only returns after the process has been terminated, and returns the status of the job. If job was already terminated before calling stop, the old state is returned without attempting termination.
+
+Could be improved by sending `SIGTERM` shortly before `SIGKILL`, but I'm aiming for simplicity here. Also separating soft and hard kills in the API level might be useful in some situations.
+
+### Status
+
+Returns job status, i.e. is it running, and the status code if the job has completed. If the job has been terminated with a signal, that is reported instead.
```

Comment:
could you list all possible statuses?
a proto enum definition would be ideal

Reply:
Sure, I'll include that to the next version. It will look something like this:

```proto
message JobStatus {
    enum Status {
        RUNNING = 0;    // Still running (value is zero)
        COMPLETED = 1;  // Completed normally (value is return code)
        SIGNAL = 2;     // Terminated by a signal (value is signal number)
        MISSING = 3;    // No such process (or access denied) (value is zero)
    }
    Status status = 1;
    int32 value = 2;
}
```

I don't think protocol buffers offer any nicer way to do this.